src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1100,16 +1142,21 @@ private static ImmutableArray<StateForCase> RemoveEvaluation(ImmutableArray<Stat
         /// <param name="other"></param>
         /// <param name="whenTrueValues">The possible values of test.Input when <paramref name="test"/> has succeeded.</param>
         /// <param name="whenFalseValues">The possible values of test.Input when <paramref name="test"/> has failed.</param>
+        /// <param name="precondition">The pre-condition that the implied status of the <paramref name="other"/> test is dependent upon.</param>
```

Comment:
Something still nags me about the meaning of `precondition`.
Would the following make sense? "`precondition` indicates when the two tests are "related" (ie. considering the same input) and so we can analyze the logical implication of `test` on `other`. If they are related in a non-trivial way (ie. not literally looking at the same DAG temp), then `sideeffect` records that the two input temps should be merged when precondition is met."
If we went with this design, then when two tests are unrelated, then we should set `precondition` to a `False` test.

Reply:
>we should set precondition to a False test.

Huh. That's exactly what was on my mind right now since I wrote `"related" being defined as "having the same input" or if P is true.` earlier. I wonder how that plays out in the rewriting. Gotta sleep on it. :)

Reply:
>then when two tests are unrelated we should set `precondition` to a `False` test.

Looking more closely, this won't make a visible difference in any case. If tests are entirely unrelated, there is no sideeffect either, so P is insignificant in the rewrite `((!P || S) && other)`.

Reply:
Thanks, that makes sense. 
I'd want to add some assertion, but we can't put assertions here about how the result of the method will be used.
I'd still recommend making that change so that this method is conceptually clear by itself (what is the meaning of `precondition` and `sideeffect`?)

Reply:
This might be confusing because there's different comments on the meaning of `precondition`, both of which are true:

`The pre-condition that the implied status of the "other" test is dependent upon`

"the implied status" includes the case where there's no implications between tests (as with other xmldocs)

`P as a pre-condition under which we need to evaluate S`

That is also true: if there is a sideeffect (defined as `An assignment node which will correspond two non-identical but related test inputs`) we want to run it only if the precondition holds.

Reply:
Tagging @333fred for his thoughts on this question (meaning of `precondition` and `sideeffect` in this API)

Reply:
Perhaps `precondition` and `sideeffect` are the wrong terms here altogether: what if we called them `relationCondition` and `relationEffect`? IE, `test` and `other` are related when `relationCondition` is true, and if it is `relationEffect` is the assignment that is executed?

Reply:
I like that. I think that's close to what I'm proposing but with better names. Just to clarify are you using the same definition of "related" as I suggested above (ie. considering the same input)?

Reply:
Judging by the name of other params and the function, I think a `decision`- prefix could work too.

`test and other are related when relationCondition is true` that's not true except if we go with False default (which tbh, I don't know how it fits into the description we have on the call-site)

Reply:
>  in case two tests are unrelated we don't simply bail

Actually, it may be a good idea to do just that (by extracting the first section of CheckConsistentDecision in another method). That's where we will have relationCondition and relationEffect or else we bail for unrelated tests. I think that would make things more clear when we're actually doing the rewrite. I'll try that out in the next iteration.